### PR TITLE
Implementation of HashJoins

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+ *                    // Row-lookup-by-hash-code can only work by the EQ operators of a join condition,
+ *                    // all other possible operators must be checked afterwards.
+ *                    emmit(combinedRow)
+ *                }
+ *            }
+ *         }
+ *     }
+ * </pre>
+ * <p>
+ * The caller of the ctor needs to pass two functions {@link #hashBuilderForLeft} and {@link #hashBuilderForRight}.
```
Comment (chaudum) [inline dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java]: ctor -> constructor

File: sql/src/test/java/io/crate/testing/UseHashJoins.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Execute "set enable_hashjoin = true" before the join statement to enable the hash join implementation.
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
+@Inherited
+public @interface UseHashJoins {
```
Comment (chaudum) [inline sql/src/test/java/io/crate/testing/UseHashJoins.java]: Can we merge this with `UseRandomizedSession` so there is only a single annotation with session context specific random parameters? Also add `UseSemiJoin` to this annotation.

Comment (chaudum) [reply sql/src/test/java/io/crate/testing/UseHashJoins.java]: Can be done in a follow up PR.

Comment (matriv) [reply sql/src/test/java/io/crate/testing/UseHashJoins.java]: @chaudum Yep make sense to merge them but in a way that we can override specific param (example semi-join) with this annotation.

**Pull request merged**